test/functional/wallet_upgradewallet.py

```diff
@@ -333,14 +341,15 @@ def copy_split_hd():
         # Check the wallet has a default key initially
         old_kvs = dump_bdb_kv(node_master_wallet)
         defaultkey = old_kvs[b'\x0adefaultkey']
-        # Upgrade the wallet. Should still have the same default key
-        wallet.upgradewallet(159900)
+        self.log.info("Upgrade the wallet. Should still have the same default key.")
+        self.test_upgradewallet(wallet, previous_version=139900, requested_version=159900, expected_version=169900)
```

Comment:
@achow101 if current version is 139900, upgradewallet 159900 is called, is the wallet actually upgrading to 169900 the expected behavior ?

Reply:
Versions less than `FEATURE_HD_SPLIT` cannot upgrade to `FEATURE_HD_SPLIT` or `FEATURE_NO_DEFAULT_KEY`. They must upgrade to `FEATURE_PRE_SPLIT_KEYPOOL` at which time both the previous features will also be applied.

If a user specifies `FEATURE_NO_DEFAULT_KEY` on ` FEATURE_HD` wallet, we should probably give an error and tell the user they can't do that.

Reply:
This test is incorrect. When `expected_version` is set, that is the version we are upgrading to. `requested_version` isn't used in that case.

Reply:
Found it. Here's a fix.

```diff
diff --git a/src/wallet/scriptpubkeyman.cpp b/src/wallet/scriptpubkeyman.cpp
index d2e1be6402..7dbbf17302 100644
--- a/src/wallet/scriptpubkeyman.cpp
+++ b/src/wallet/scriptpubkeyman.cpp
@@ -453,7 +453,7 @@ bool LegacyScriptPubKeyMan::Upgrade(int prev_version, int new_version, bilingual
         hd_upgrade = true;
     }
     // Upgrade to HD chain split if necessary
-    if (IsFeatureSupported(new_version, FEATURE_HD_SPLIT)) {
+    if (!IsFeatureSupported(prev_version, FEATURE_HD_SPLIT) && IsFeatureSupported(new_version, FEATURE_HD_SPLIT)) {
         WalletLogPrintf("Upgrading wallet to use HD chain split\n");
         m_storage.SetMinVersion(FEATURE_PRE_SPLIT_KEYPOOL);
         split_upgrade = FEATURE_HD_SPLIT > prev_version;
diff --git a/test/functional/wallet_upgradewallet.py b/test/functional/wallet_upgradewallet.py
index 7cae913fa1..fdd52ba574 100755
--- a/test/functional/wallet_upgradewallet.py
+++ b/test/functional/wallet_upgradewallet.py
@@ -342,7 +342,7 @@ class UpgradeWalletTest(BitcoinTestFramework):
         old_kvs = dump_bdb_kv(node_master_wallet)
         defaultkey = old_kvs[b'\x0adefaultkey']
         self.log.info("Upgrade the wallet. Should still have the same default key.")
-        self.test_upgradewallet(wallet, previous_version=139900, requested_version=159900, expected_version=169900)
+        self.test_upgradewallet(wallet, previous_version=139900, requested_version=159900, expected_version=159900)
         new_kvs = dump_bdb_kv(node_master_wallet)
         up_defaultkey = new_kvs[b'\x0adefaultkey']
         assert_equal(defaultkey, up_defaultkey)
```